conf/db.c

```diff
@@ -142,63 +142,55 @@ struct confx_ctx {
 	struct m0_xcode_ctx     c_xctx;
 };
 
-static m0_bcount_t confdb_ksize(const void *key)
-{
-	return sizeof(struct m0_fid);
-}
-
-static m0_bcount_t confdb_vsize(const void *val)
-{
-	return m0_confx_sizeof() + sizeof(struct confx_allocator);
-}
-
-static const struct m0_be_btree_kv_ops confdb_ops = {
-	.ko_type    = M0_BBT_CONFDB,
-	.ko_ksize   = &confdb_ksize,
-	.ko_vsize   = &confdb_vsize,
-	.ko_compare = (void *)&m0_fid_cmp
-};
-
 /* ------------------------------------------------------------------
  * Tables
  * ------------------------------------------------------------------ */
 
-static const char btree_name[] = "conf";
+static const char     btree_name[] = "conf";
+static const uint32_t rnode_sz     = 4096;
 
 static int confdb_table_init(struct m0_be_seg     *seg,
-			     struct m0_be_btree  **btree,
+			     struct m0_btree      *btree,
 			     struct m0_be_tx      *tx,
 			     const struct m0_fid  *btree_fid)
 {
-	int rc;
+	uint8_t              *rnode;
+	struct m0_btree_op    b_op = {};
+	struct m0_btree_type  bt;
+	uint32_t              rnode_sz_shift;
+	int                   rc;
 
 	M0_ENTRY();
 
-	M0_BE_ALLOC_PTR_SYNC(*btree, seg, tx);
-	m0_be_btree_init(*btree, seg, &confdb_ops);
-	M0_BE_OP_SYNC(op, m0_be_btree_create(*btree, tx, &op, btree_fid));
-
-	rc = m0_be_seg_dict_insert(seg, tx, btree_name, *btree);
-	if (rc == 0 && M0_FI_ENABLED("ut_confdb_create_failure"))
-		rc = -EINVAL;
-	if (rc != 0) {
-		confdb_table_fini(seg);
-		m0_confdb_destroy(seg, tx);
+	M0_ASSERT(rnode_sz > 0 && m0_is_po2(rnode_sz));
+	rnode_sz_shift = __builtin_ffsl(rnode_sz) - 1;
+	M0_BE_ALLOC_ALIGN_ARR_SYNC(rnode, rnode_sz, rnode_sz_shift, seg, tx);
+	bt = (struct m0_btree_type){ .tt_id = M0_BT_CONFDB,
+				     .ksize = sizeof (struct m0_fid),
+				     .vsize = -1,
+				   };
+	rc = M0_BTREE_OP_SYNC_WITH_RC(&b_op, m0_btree_create(rnode, rnode_sz,
+							     &bt, &b_op, btree,
+							     seg, btree_fid,
+							     tx));
+	if (rc != 0)
+		M0_BE_FREE_ALIGN_ARR_SYNC(rnode, seg, tx);
+	else {
+		rc = m0_be_seg_dict_insert(seg, tx, btree_name, rnode);
+		if (rc == 0 && M0_FI_ENABLED("ut_confdb_create_failure"))
+			rc = -EINVAL;
+		if (rc != 0) {
+			M0_BTREE_OP_SYNC_WITH_RC(&b_op,
```

Comment:
Add check for return code.

Reply:
This is done on purpose since because we need to return the previous rc value to the caller. If have a failure for closing the tree that is secondary.